### PR TITLE
Speed up `bundler/setup` time

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
## What was the end-user or developer problem that led to this PR?

We should make `bundler/setup` as fast as possible.

## What is your fix for the problem, implemented in this PR?

Avoid unnecessary work of figuring out which specs should be kept locked from the lockfile and which ones should be discarded for resolving, whenever possible (in frozen mode, or when no changes detected).

This speeds up `bundler/setup` time of `rails/rails` Gemfile on my machine by about 30%.

```
✗ hyperfine 'BUNDLER_VERSION=2.3.12 ruby -rbundler/setup -e1' 'BUNDLER_VERSION=2.4.0.dev ruby -rbundler/setup -e1'
Benchmark 1: BUNDLER_VERSION=2.3.12 ruby -rbundler/setup -e1
  Time (mean ± σ):     368.9 ms ±  10.7 ms    [User: 309.4 ms, System: 46.9 ms]
  Range (min … max):   360.4 ms … 395.4 ms    10 runs
 
Benchmark 2: BUNDLER_VERSION=2.4.0.dev ruby -rbundler/setup -e1
  Time (mean ± σ):     267.3 ms ±   3.2 ms    [User: 212.3 ms, System: 44.8 ms]
  Range (min … max):   263.9 ms … 272.1 ms    10 runs
 
Summary
  'BUNDLER_VERSION=2.4.0.dev ruby -rbundler/setup -e1' ran
    1.38 ± 0.04 times faster than 'BUNDLER_VERSION=2.3.12 ruby -rbundler/setup -e1'
```

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [ ] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

**Comment by deivid-rodriguez** [comment]
There are big speed ups on `rails/rails` repository Gemfile because it's big, I suspect it should also be the case for @ngan's. But this is also noticiable (5-10% speed up) on a default Rails generated application 🎉:

```
hyperfine 'BUNDLER_VERSION=2.3.12 ruby -rbundler/setup -e1' 'BUNDLER_VERSION=2.4.0.dev ruby -rbundler/setup -e1'
Benchmark 1: BUNDLER_VERSION=2.3.12 ruby -rbundler/setup -e1
  Time (mean ± σ):     156.8 ms ±   1.3 ms    [User: 114.2 ms, System: 34.0 ms]
  Range (min … max):   155.2 ms … 161.1 ms    18 runs
 
Benchmark 2: BUNDLER_VERSION=2.4.0.dev ruby -rbundler/setup -e1
  Time (mean ± σ):     147.5 ms ±   0.7 ms    [User: 105.7 ms, System: 33.2 ms]
  Range (min … max):   146.5 ms … 149.4 ms    19 runs
 
Summary
  'BUNDLER_VERSION=2.4.0.dev ruby -rbundler/setup -e1' ran
    1.06 ± 0.01 times faster than 'BUNDLER_VERSION=2.3.12 ruby -rbundler/setup -e1'
```

**Comment by ngan** [comment]
Nice work, @deivid-rodriguez!

I ran some benchmarks for our gemfile.

`master`:
```
$ time bin/rubocop --version
bin/rubocop --version  2.63s user 0.40s system 97% cpu 3.115 total
```

`speed-up-bundler-setup`:
```
$ time bin/rubocop --version
bin/rubocop --version  1.96s user 0.37s system 93% cpu 2.486 total
```

So 25% improvement! It's good...but it's unfortunately still [not as good as subbundle](https://github.com/bigrails/subbundle#anecdotal-benchmarks).

With subbundle:
```
$ time bin/rubocop --version
bin/rubocop --version  0.51s user 0.16s system 92% cpu 0.720 total
```

**Comment by deivid-rodriguez** [comment]
Thanks for testing this out @ngan!

Your results are along the lines of what I expected. I think this effectively closes https://github.com/rubygems/rubygems/issues/5345, right?

Regarding the rest of the speed up, I think it probably can be obtained through an `only` option now? I recall the implementation of an `only` option was very simple, I think someone proposed it at some point. It would be good to get that rebased and ready, and perform some experiments with it to see how much it buys us.

In any case, I have some extra speed ups in mind that I can hopefully implement and will get to every user "for free".

**Comment by deivid-rodriguez** [comment]
@ngan Here's an extra patch that you can try out to see where it gets us:

```diff
diff --git a/bundler/lib/bundler/definition.rb b/bundler/lib/bundler/definition.rb
index 5c659608b0..ff4004fc11 100644
--- a/bundler/lib/bundler/definition.rb
+++ b/bundler/lib/bundler/definition.rb
@@ -146,7 +146,7 @@ def initialize(lockfile, dependencies, sources, unlock, ruby_version = nil, opti
       @dependency_changes = converge_dependencies
       @local_changes = converge_locals
 
-      @locked_specs_incomplete_for_platform = !@locked_specs.for(requested_dependencies & expand_dependencies(locked_dependencies), true, true)
+      @locked_specs_incomplete_for_platform = !@locked_specs.for(requested_dependencies & locked_dependencies, true, true)
 
       @requires = compute_requires
     end
@@ -245,10 +245,9 @@ def specs_for(groups)
 
     def dependencies_for(groups)
       groups.map!(&:to_sym)
-      deps = current_dependencies.reject do |d|
+      current_dependencies.reject do |d|
         (d.groups & groups).empty?
       end
-      expand_dependencies(deps)
     end
 
     # Resolve all the dependencies specified in Gemfile. It ensures that
```

**Comment by duckinator** [comment]
@deivid-rodriguez This looks really good, and is a surprisingly minor change for a noticeable speed improvement!

Are there any known/expected changes in behavior? Or things that need tests to be added? (None jumped out at me when I looked.)

**Comment by ngan** [comment]
Hey @deivid-rodriguez Im on vacation and won’t have a computer until Monday. Will test this asap when I get back. I’m excited about this because if I recall correctly, expand_dependencies was the culprit…

**Comment by deivid-rodriguez** [comment]
@duckinator No, I don't think so, although I'm preparing other patches which in my testing improve performance of setting up `rails/rails` repository's Gemfile by about 100%. And those might actually bring a small change.

@ngan Yes, I'm expecting that too, given that you have a lot of dependencies, and `expand_dependencies` loops through all of them.

**Comment by ngan** [comment]
@deivid-rodriguez I applied [the second patch](https://github.com/rubygems/rubygems/pull/5503#issuecomment-1113106680) on top of the first. Unfortunately, it didn't yield any improvements.

```
$ time bin/rubocop --version
bin/rubocop --version  2.00s user 0.38s system 95% cpu 2.477 total
```

**Comment by deivid-rodriguez** [comment]
Too bad! Anyways, I'll merge the current speed up and iterate on this in future versions!

**Comment by ngan** [comment]
Oh definitely! I’ll find some time this week and see if I can help out.
